AndroidSDKCore/src/main/java/com/leanplum/callbacks/MessageDisplayedCallback.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2013, Leanplum, Inc. All rights reserved.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.leanplum.callbacks;
+
+import com.leanplum.ActionContext;
+
+import java.util.Date;
+
+/**
+ * Message displayed callback.
+ *
+ * @author Mayank Sanganeria
+ */
+public abstract class MessageDisplayedCallback implements Runnable {
+
+  private String messageID;
+  private String messageBody;
+  private String recipientUserID;
+  private Date deliveryDateTime;
+
+  public void setMessageBody(String messageBody) {
+    this.messageBody = messageBody;
+  }
+
+  public void setMessageID(String messageID) {
+    this.messageID = messageID;
+  }
+
+  public void setRecipientUserID(String recipientUserID) {
+    this.recipientUserID = recipientUserID;
+  }
+
+  public void setDeliveryDateTime(Date deliveryDateTime) {
+    this.deliveryDateTime = deliveryDateTime;
+  }
+
+  public void run() {
+    this.messageDisplayed(messageID, messageBody, recipientUserID, deliveryDateTime);
+  }
+
+  public abstract void messageDisplayed(String messageID, String messageBody,
```

Comment:
I don't want to stop this up with scope creep again, but if we really think about our customer here this could be more useful. They want a blob of this information that they will store somewhere. If we had a POJO representing the "message" with all the required fields, it would be scalable for the future and give them an easy way to deal with it/JSON-ify it. Now they have to hard-code creating their own object into this function signature that would require a code change from them to update.

Also FYI, it's always a red flag to have many arguments in a function signature.